libparsec/crates/platform_device_loader/src/web/error.rs

```diff
@@ -0,0 +1,205 @@
+    GetItemStorageError = {
+        #[display("Failed to get item {key} from storage ({error:?})")]
+        GetItemStorage {
+            key: String,
+            error: JsValue
+        }
+    };
+    SetItemStorageError = {
+        #[display("Failed to set item {key} to storage ({error:?})")]
+        SetItemStorage {
+            key: String,
+            error: JsValue
+        }
+    };
+    RemoveItemStorageError = {
+        #[display("Failed to remove item {key} from storage ({error:?})")]
+        RemoveItemStorage {
+            key: String,
+            error: JsValue
+        }
+    };
+    JsonDeserializationError = {
+        JsonDecode(serde_json::Error)
+    };
+    JsonSerError = {
+        JsonEncode(serde_json::Error)
+    };
+    RmpDecodeError = {
+        RmpDecode(libparsec_types::RmpDecodeError)
+    };
+    Base64DecodeError = {
+        B64Decode(base64::DecodeError)
+    };
+    ListAvailableDevicesError = GetItemStorageError
+        || JsonDeserializationError
+        || InvalidPathError
+        || LoadAvailableDeviceError;
+    GetRawDeviceError = GetItemStorageError
+        || Base64DecodeError
+        || {
+        #[display("No device for '{key}'")]
+        Missing {
+            key: String
+        },
+    };
+    LoadAvailableDeviceError = GetRawDeviceError || RmpDecodeError;
+    InvalidPathError = {
+        #[display("Invalid path {}", path.display())]
+        InvalidPath {
+            path: std::path::PathBuf
+        }
+    };
+    LoadDeviceError = InvalidPathError
+        || GetRawDeviceError
+        || RmpDecodeError
+        || {
+            InvalidFileType,
+            GetSecretKey(libparsec_crypto::CryptoError),
+            DecryptAndLoad(crate::DecryptDeviceFileError),
+        };
+    SaveDeviceError = SaveDeviceFileError || InvalidPathError;
+    SaveDeviceFileError = SetItemStorageError || AddItemToListError;
+    AddItemToListError = GetRawDeviceError
+        || SetItemStorageError
+        || JsonDeserializationError
+        || JsonSerError;
+    RemoveItemFromListError = GetRawDeviceError
+        || SetItemStorageError
+        || JsonDeserializationError
+        || JsonSerError;
+    ArchiveDeviceError = GetItemStorageError
+        || AddItemToListError
+        || RemoveItemFromListError
+        || JsonDeserializationError
+        || InvalidPathError;
+    RemoveDeviceError = GetItemStorageError
+        || RemoveItemFromListError
+        || RemoveItemStorageError
+        || JsonDeserializationError
+        || InvalidPathError;
+}
+
+impl From<LoadDeviceError> for crate::LoadDeviceError {
+    fn from(value: LoadDeviceError) -> Self {
+        match value {
+            LoadDeviceError::InvalidFileType => Self::InvalidData,
+            LoadDeviceError::GetSecretKey(_) => Self::DecryptionFailed,
+            LoadDeviceError::DecryptAndLoad(e) => e.into(),
+            LoadDeviceError::InvalidPath { .. } => Self::InvalidPath(anyhow::anyhow!("{value}")),
```

Comment:
In web, the path is totally under our control:
- in prod it is a dummy value set at compilation time
- in test, we are generating this value to whatever suits us

So I think it is safe to assume the path is always valid utf8, and just panic if that's not the case (we should just have a comment next to panic to explain why it should never occur in practice)

Reply:
done